### PR TITLE
Added prettier, lint-staged and husky.

### PR DESCRIPTION
**Pull request by lukester1975**
As suggested in #68.

Just a stab at a prettier config. (I personally don't understand the preference for 80 columns... my monitor is wider than I am tall, and I'm 6'5")

print-width 100 example: https://github.com/lukester1975/vscode-meson/compare/78052ee5f2c33fcdaa25b93698726b10d658a71d..1fbd08082f804f73366ba032005d3101005d9f1d
80: https://github.com/lukester1975/vscode-meson/compare/78052ee5f2c33fcdaa25b93698726b10d658a71d..856fbf0fa785f62117d4af8da0ad58baa0fc2645

### CODE REVIEW

File: .husky/pre-commit [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
```
Comment (tristan957) [inline .husky/pre-commit:1]: `/bin/sh` is the preferred shebang I believe

File: .prettierignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.git
+node_modules
+out
```
Comment (tristan957) [inline .prettierignore:3]:
```ignore
/.git/
/node_modules/
/out/
```

File: package.json [hunk 3]
```diff
@@ -319,14 +319,21 @@
...
     "@types/vscode": "^1.1.59",
+    "husky": "^8.0.3",
+    "lint-staged": "^13.2.1",
+    "prettier": "^2.8.7",
     "typescript": "^4.4.4"
   },
-  "dependencies": {}
+  "dependencies": {},
+  "prettier": {
+    "printWidth": 100
```
Comment (tristan957) [inline package.json:337]: Let's just finish the rest of the config.

Determine if we are using tabs or spaces, and choose it.

semi: true
singleQuote: false
quoteProps: as-needed
jsxSingleQuote: false
trailingCommas: all
bracketSpacing: true
bracketSameLine: false
arrowParens: true
wrapProse: always
endOfLine: lf

**Review by tristan957 — COMMENTED** [review]
Can you add a GitHub Action for using Prettier?

Here is an example: https://github.com/tristan957/libmerr/blob/master/.github/workflows/prettier.yaml

File: .husky/pre-commit (hunk 1 above)
Comment (lukester1975) [reply .husky/pre-commit:1]: I wouldn't want to be on that jury...! However, husky creates the hooks like this (i.e. with `husky add pre-commit "yarn --silent lint-staged -r"`).

**Comment by lukester1975** [comment]
GH action - sure, will take a look at some point.

**Comment by dcbaker** [comment]
I haven't looked at anything but the initial MR message. I'd be happy with 100 or 120 column widths. I'd prefer not to go wider than that as I usually have two files opened side-by side. I think Meson itself is 100?

File: package.json (hunk 3 above)
Comment (dcbaker) [reply package.json]: spaces

**Comment by tristan957** [comment]
@lukester1975 any time to come back to this?

**Comment by lukester1975** [comment]
I'm not a GH actions person but this appears to work based of yours @tristan957.

* meson uses a line length of 120 by the looks of it (in `.flake8`), so did the same here.
* Since there's an `.editorconfig`, I put the line length there as prettier will use that.

Not sure I see the point of specifying options with the default values: as soon as a new value is added, it's no longer complete anyway. Plus if a default ever changes, highlighting that fact and reviewing what is preferred is a good thing IMHO. Plus package.json is too long as it is...

File: .github/workflows/prettier.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+# SPDX-FileCopyrightText: 2023 Tristan Partin <tristan@partin.io>
```
Comment (tristan957) [inline .github/workflows/prettier.yml:3]: TBH, you can just drop this. It isn't important to me

File: .github/workflows/prettier.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+name: Prettier
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+    paths:
+      - README.md
```
Comment (tristan957) [inline .github/workflows/prettier.yml:11]:
```
- **.json
- **.md
- **.ts
- **.yml
```

File: .github/workflows/prettier.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Setup node
+        uses: actions/setup-node@v3
+        with:
+          node-version: latest
+
+      - name: Checkout
+        uses: actions/checkout@v3
```
Comment (tristan957) [inline .github/workflows/prettier.yml:31]: You can bump this to v4.

File: .github/workflows/prettier.yml (hunk 5 above)
Comment (lukester1975) [reply .github/workflows/prettier.yml]: Heh, silly me.

That said, given the `prettier:check` script specifies the patterns, might as well not filter here? Not like PRs are flying in so even if there's a perf hit surely not an issue...

File: .github/workflows/codeql-analysis.yml [hunk 7]
```diff
@@ -16,12 +16,12 @@ permissions:
```
[the remark this answers is not in the document]
Comment (lukester1975) [reply .github/workflows/codeql-analysis.yml:16]: Already is one apparently, so I've added the commit. Will try to remember it on next rebase, unless this is getting merged soon...

File: .github/workflows/prettier.yml (hunk 5 above)
Comment (tristan957) [reply .github/workflows/prettier.yml]: That is fine. Can always change it later.

File: .github/workflows/codeql-analysis.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (tristan957) [reply .github/workflows/codeql-analysis.yml:16]: I am ready to merge. Are you good to go?

**Comment by tristan957** [comment]
Weird. I will watch it.

**Pull request merged**